components/brave_shields/adblock/rs/src/lib.rs

```diff
@@ -0,0 +1,198 @@
+    fn hidden_class_id_selectors(
+      &self,
+      classes: &CxxVector<CxxString>,
+      ids: &CxxVector<CxxString>,
+      exceptions: &CxxVector<CxxString>,
+    ) -> VecStringResult;
+    /// Returns the blocker debug info containing regex info.
+    fn get_debug_info(&self) -> BlockerDebugInfo;
+    /// Removes a regex entry by the id.
+    fn discard_regex(&mut self, regex_id: u64);
+    /// Sets a discard policy for the regex manager.
+    fn set_regex_discard_policy(&mut self, new_discard_policy: &RegexManagerDiscardPolicy);
+  }
+
+  unsafe extern "C++" {
+    include!("brave/components/brave_shields/adblock/rs/src/adblock_domain_resolver.h");
+
+    /// Wrapper function for net::registry_controlled_domains::GetDomainAndRegistry
+    /// implemented in Chromium.
+    fn resolve_domain_position(host: &CxxString) -> DomainPosition;
+  }
+
+  struct DomainPosition {
+    start: u32,
+    end: u32,
+  }
+
+  #[derive(Default)]
+  struct BlockerResult {
+    matched: bool,
+    important: bool,
+    has_exception: bool,
+    redirect: OptionalString,
+    rewritten_url: OptionalString,
+  }
+
+  struct RegexDebugEntry {
+    id: u64,
+    regex: OptionalString,
+    unused_secs: u64,
+    usage_count: usize,
+  }
+
+  struct BlockerDebugInfo {
+    regex_data: Vec<RegexDebugEntry>,
+    compiled_regex_count: usize,
+  }
+
+  struct RegexManagerDiscardPolicy {
+    cleanup_interval_secs: u64,
+    discard_unused_secs: u64,
+  }
+
+  #[derive(Default)]
+  struct FilterListMetadata {
+    homepage: OptionalString,
+    title: OptionalString,
+  }
+
+  enum ResultKind {
+    Success,
+    JsonError,
+    Utf8Error,
+    AdblockError,
+  }
+
+  // The following structs are not generic because generic type parameters are
+  // not yet supported in cxx.
+  // Created custom Result structs because cxx auto converts Result<T> to std::exception, and
+  // exceptions are not allowed in Chromium.
+
+  // To check a result for success/error, evaluate the result_kind == ResultKind::Success.
+  // If the condition is false, the result_kind will contain an enum value describing the type of error,
+  // and error_message may contain further details about the error.
+  struct EmptyTupleResult {
+    result_kind: ResultKind,
+    error_message: String,
+  }
+
+  struct StringResult {
+    value: String,
+    result_kind: ResultKind,
+    error_message: String,
+  }
+
+  struct VecStringResult {
+    value: Vec<String>,
+    result_kind: ResultKind,
+    error_message: String,
+  }
+
+  struct BoolResult {
+    value: bool,
+    result_kind: ResultKind,
+    error_message: String,
+  }
+
+  struct BoxEngineResult {
+    value: Box<Engine>,
+    result_kind: ResultKind,
+    error_message: String,
+  }
+
+  struct FilterListMetadataResult {
+    value: FilterListMetadata,
+    result_kind: ResultKind,
+    error_message: String,
+  }
+
+  // TODO(djandries): Come up with a better name?
+  struct BlockerResultResult {
```

Comment:
thanks for the insight, that's a fair assessment IMO.

I decided to add handling for this errors because I was unsure if it would be okay to panic in rust land. A search on Slack shows discussions, but nothing I could deem concrete. Also saw a message from @iefremov in similar threads stating he was having issues with UTF-8 related panics in various crates, but I'm  unsure if adblock was relevant in that discussion.

If you and/or others can confirm that:

a) it is permissible to panic in Rust land in the event of an unrecoverable error, and
b) a UTF-8 error in the adblock context can be deemed unrecoverable (which I think you implied already, but still wanted to confirm anyway)

then I will be happy to simplify most functions by removing the `Utf8Error` from the `ResultKind` and returning values directly without using result types in any functions where only UTF-8 errors can occur.

Reply:
You can see https://github.com/brave/adblock-rust/pull/264#issuecomment-1454242209 for my thoughts on this - we do already have a few `unwrap` calls in `adblock-rust`, all annotated with a comment explaining what invariant is being upheld to ensure the code will run safely.

I'd much rather be able to catch and fix the panics in Nightly/Beta (backtrace handles them just fine) than have unknown behavior floating around in users' installations. Especially when it comes to something as critical as UTF-8 validity - that's a likely target for security vulnerabilities in parsing code.

Reply:
> removing the `Utf8Error` from the `ResultKind`, and returning values directly without using result types in any functions where only UTF-8 errors can occur

I will note though, we definitely shouldn't apply this to _every_ API. For example, in the case of custom subscriptions, filter lists downloaded from arbitrary URLs are passed directly to `adblock-rust`. It'd be rare since we forbid files that are served with a mimetype other than plaintext, but it's totally possible for a server to send a malformed text file, so we should provide a way to recover from that.